CRM/MembershipExtras/Hook/Pre/ContributionRecur.php

```diff
@@ -0,0 +1,158 @@
+    if ($this->operation == 'edit' && $this->isManualPaymentPlan()) {
+      $this->rectifyPaymentPlanStatus();
+    }
+  }
+
+  /**
+   * Calculates recurring contribution status.
+   */
+  private function rectifyPaymentPlanStatus() {
+    $status = $this->calculateRecurringContributionStatus();
+
+    $contributionStatus = CRM_Contribute_PseudoConstant::contributionStatus(NULL, 'name');
+    $statusID = array_search($status, $contributionStatus);
+    $this->params['contribution_status_id'] = $statusID;
+
+    if ($status === 'Completed' && $this->recurringContribution['installments'] > 1) {
+      $this->params['end_date'] = $this->generateNewPaymentPlanEndDate();
+    }
+  }
+
+  /**
+   * Generates end date for recurring contribution from last paid contribution.
+   *
+   * @return string
+   */
+  private function generateNewPaymentPlanEndDate() {
+    $lastPaymentPlanContribution = civicrm_api3('Contribution', 'get', [
+      'sequential' => 1,
+      'return' => ['receive_date'],
+      'contribution_recur_id' => $this->recurringContribution['id'],
+      'options' => ['sort' => 'id DESC', 'limit' => 1],
+    ]);
+
+    $endDate = NULL;
+    if (!empty($lastPaymentPlanContribution['values'][0]['receive_date'])) {
+      $endDate = $lastPaymentPlanContribution['values'][0]['receive_date'];
+    }
+
+    return $endDate;
+  }
+
+  /**
+   * Calculates the status of the current recurring contribution.
+   *
+   * @return string
+   */
+  private function calculateRecurringContributionStatus() {
+    $paidInstallmentsCount = civicrm_api3('Contribution', 'getcount', [
+      'contribution_recur_id' => $this->recurringContribution['id'],
+      'contribution_status_id' => 'Completed',
+    ]);
+
+    $partiallyPaidInstallmentsCount = civicrm_api3('Contribution', 'getcount', [
+      'contribution_recur_id' => $this->recurringContribution['id'],
+      'contribution_status_id' => 'Partially paid',
+    ]);
+
+    $allPaid = $paidInstallmentsCount >= $this->recurringContribution['installments'];
+    $moreThanOneInstallment = $this->recurringContribution['installments'] > 1;
+
+    switch (true) {
+      case $moreThanOneInstallment && $allPaid:
+        $status = 'Completed';
+        break;
+
+      case $paidInstallmentsCount >= 1 || $partiallyPaidInstallmentsCount >=  1:
+        $status = 'In Progress';
+        break;
+
+      default:
+        $status = 'Pending';
+    }
+
+    return $status;
+  }
+
+  /**
+   * Checks if current recurring contribution corresponds to a manual payment
+   * plan.
+   */
+  private function isManualPaymentPlan() {
+    $payLaterProcessorID = 0;
+    $manualPaymentProcessorsIDs = array_merge([$payLaterProcessorID], CRM_MembershipExtras_Service_ManualPaymentProcessors::getIDs());
+    $isManualPaymentProcessor = in_array($this->recurringContribution['payment_processor_id'], $manualPaymentProcessorsIDs);
+
+    if ($isManualPaymentProcessor) {
+      return TRUE;
+    }
+
+    return FALSE;
+  }
+
+}
```

Comment:
Blank line at EOF

Reply:
It's there, it just looks like its's not. When tehre is no line at EOF, github shows it like this:

![image](https://user-images.githubusercontent.com/21999940/48414539-0cd0a880-e719-11e8-9ae1-804255616a2e.png)